### ISSUE TITLE
glob exclude patterns inconsistency

### ISSUE THREAD

**Issue by hnridder** [thought-machine/please#874]
Initialize a new repo and create (touch) the following files:
```
src/a.not
src/a.txt
src/b.txt
src/c.txt
src/d/a.txt
src/d/b.txt
src/d/c.txt
src/d/e/a.txt
```
Create a BUILD file with the following contents:
```
filegroup(
   name = 'one',
   srcs = glob(['src/**'], exclude=['src/d/*']),
)
```
Create a directory java, copy the entire src tree there and the BUILD file as well (actually I created a symlink).
## Difference in :one and //java:one
```
$ plz query print :one //java:one
# //:one:
filegroup(
    name = 'one',
    srcs = [
        'src/a.not',
        'src/a.txt',
        'src/b.txt',
        'src/c.txt',
    ],
    timeout = 600,
)

# //java:one:
filegroup(
    name = 'one',
    srcs = [
        'src/a.not',
        'src/a.txt',
        'src/b.txt',
        'src/c.txt',
        'src/d/e/a.txt',
    ],
    timeout = 600,
)
```
I expect the same set of filenames in both cases.

## Multiple exclude patterns
Change the exclude list to `['*.not', 'src/d/*']`. Now the output is:
```
# //:one:
filegroup(
    name = 'one',
    srcs = [
        'src/a.txt',
        'src/b.txt',
        'src/c.txt',
        'src/d/a.txt',
        'src/d/b.txt',
        'src/d/c.txt',
        'src/d/e/a.txt',
    ],
    timeout = 600,
)

# //java:one:
filegroup(
    name = 'one',
    srcs = [
        'src/a.txt',
        'src/b.txt',
        'src/c.txt',
        'src/d/a.txt',
        'src/d/b.txt',
        'src/d/c.txt',
        'src/d/e/a.txt',
    ],
    timeout = 600,
)
```
It looks as if the second exclude pattern is ignored. However, if we change the order of the exclude list to `['src/d/*', '*.not']` then the output becomes
```
# //:one:
filegroup(
    name = 'one',
    srcs = [
        'src/a.txt',
        'src/b.txt',
        'src/c.txt',
    ],
    timeout = 600,
)

# //java:one:
filegroup(
    name = 'one',
    srcs = [
        'src/a.txt',
        'src/b.txt',
        'src/c.txt',
        'src/d/e/a.txt',
    ],
    timeout = 600,
)
```

## How to exclude an entire directory tree
As background information, I'm trying to exclude multiple directory trees and it would be nice if please would support this, which I don't think it does at the moment. However, due to the inconsistencies mentioned above, I'm not entirely certain what please does do and what it doesn't do.

I would like to take all files rooted at T, _except_ those that are contained in a subdirectory (or deeper) of T/I/A. Files in T/I/A themselves should be included. So something like `glob(['T/**'], exclude=['T/I/A/*/**'])`.

It would be nice if please could support this, e.g. by allowing \*\* in exclude patterns, or by treating an exclude pattern ending in / as a subtree match.

**Comment by Tatskaari** [comment]
Hi @hnridder, thank you for the excellent bug report! I was able to replicate this issue and have fixed it in #927

**Comment by Tatskaari** [comment]
Had a few issues with that library keeping files open too long. #933 fixes our implementation of glob.
